src/cfg/cfg-traversal.h

```diff
@@ -210,41 +219,89 @@ struct CFGWalker : public ControlFlowWalker<SubType, VisitorType> {
+        if (self->unwindExprStack[i]->template cast<Try>()->hasCatchAll()) {
+          break;
+        }
+        for (auto* block : self->unwindCatchStack[i - 1]) {
+          self->link(last, block);
+        }
+      }
     }
   }
 
   static void doStartTry(SubType* self, Expression** currp) {
+    auto* curr = (*currp)->cast<Try>();
     auto* last = self->currBasicBlock;
-    self->startBasicBlock(); // catch body's first block
-    self->catchStack.push_back(self->currBasicBlock);
+    self->unwindCatchStack.emplace_back();
+    self->unwindExprStack.push_back(curr);
+    for (Index i = 0; i < curr->catchBodies.size(); i++) {
+      self->startBasicBlock(); // catch body's first block
+      self->unwindCatchStack.back().push_back(self->currBasicBlock);
+    }
     self->currBasicBlock = last; // reset to the current block
   }
 
-  static void doStartCatch(SubType* self, Expression** currp) {
+  static void doStartCatches(SubType* self, Expression** currp) {
     self->tryStack.push_back(self->currBasicBlock); // last block of try body
-    self->currBasicBlock = self->catchStack.back();
-    self->catchStack.pop_back();
+    self->processCatchStack.push_back(self->unwindCatchStack.back());
+    self->unwindCatchStack.pop_back();
+    self->unwindExprStack.pop_back();
+  }
+
+  static void doStartCatch(SubType* self, Expression** currp, Index i) {
+    self->currBasicBlock = self->processCatchStack.back()[i];
+  }
+
+  static void doEndCatch(SubType* self, Expression** currp, Index i) {
+    self->processCatchStack.back()[i] = self->currBasicBlock;
   }
 
   static void doEndTry(SubType* self, Expression** currp) {
-    auto* last = self->currBasicBlock;
     self->startBasicBlock(); // continuation block after try-catch
-    // catch body's last block -> continuation block
-    self->link(last, self->currBasicBlock);
+    // each catch body's last block -> continuation block
+    for (auto* last : self->processCatchStack.back()) {
+      self->link(last, self->currBasicBlock);
+    }
     // try body's last block -> continuation block
     self->link(self->tryStack.back(), self->currBasicBlock);
     self->tryStack.pop_back();
+    self->processCatchStack.pop_back();
   }
 
   static void doEndThrow(SubType* self, Expression** currp) {
-    // We unwind to the innermost catch, if any
-    if (!self->catchStack.empty()) {
-      self->link(self->currBasicBlock, self->catchStack.back());
+    // We unwind to the innermost catches, if any
+    if (!self->unwindCatchStack.empty()) {
+      for (auto* block : self->unwindCatchStack.back()) {
+        self->link(self->currBasicBlock, block);
+      }
+      // Exception thrown. Create a link to each catch within the innermost try.
+      for (auto* block : self->unwindCatchStack.back()) {
+        self->link(self->currBasicBlock, block);
+      }
+      // Until we encounter try-catch_all, an exception can unwind to outer
+      // catches
```

Comment:
this pattern appears in `doEndCall` as well, I think - maybe it can be refactored to a shared place?

Reply:
Extracted the common part as `doEndThrowingInst`.